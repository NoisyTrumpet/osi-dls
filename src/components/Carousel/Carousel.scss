/** 
 * GliderJS Styling Override
 */
.glider-dot:hover,
.glider-dot:focus,
.glider-dot.active {
  background-color: map-get($colors, 'primary1');
}

.glider-dot {
  background: #d6d6df;
  width: $space-8;
  height: $space-8;
  margin: $space-16;
}

.glider-dot.active {
  background-color: map-get($colors, 'primary1');
}

// https://www.w3schools.com/howto/howto_css_hide_scrollbars.asp
.glider,
.glider-track {
  scrollbar-width: none; /* Firefox */
  -ms-overflow-style: none; /* IE 10+ */
  &::-webkit-scrollbar {
    display: none; /* Chrome/Safari/Webkit */
  }
}

/** 
 * Carousel Styling
 */
.carousel {
  width: 100%;
  height: 100%;
  .carousel__track-container {
    width: 100%;
    height: 100%;
    background: map-get($colors, 'brand-transparent');
    .carousel__track {
      height: 100%;
      .list-item {
        background-color: map-get($colors, 'brand-white');
      }
    }
  }
}

// CSS for a carousel with 1:1 aspect ratio
.carousel-square {
  .carousel__track-container {
    padding-top: 100%; /* 1:1 Aspect Ratio */
    position: relative;
    .carousel__track {
      position: absolute;
      top: 0;
      left: 0;
      bottom: 0;
      right: 0;
    }
  }
}
