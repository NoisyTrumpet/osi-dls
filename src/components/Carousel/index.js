export { default as Carousel } from './Carousel'; // eslint-disable-line import/prefer-default-export
