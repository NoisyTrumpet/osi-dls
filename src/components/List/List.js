import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import ListItem from './ListItem';
import './List.scss';

const List = ({ children, className, addClass, ordered, horizontal, ...rest }) => {
  const classes = classNames(
    className,
    addClass,
    { 'list-horizontal': horizontal },
  );
  const ListTag = ordered ? 'ol' : 'ul';

  return (
    <ListTag
      className={classes}
      {...rest} // eslint-disable-line react/jsx-props-no-spreading
    >
      {children}
    </ListTag>
  );
};

List.Item = ListItem;

List.defaultProps = {};

List.propTypes = {
  ordered: PropTypes.bool,
  horizontal: PropTypes.bool,
  children: PropTypes.node,
  className: PropTypes.oneOf([
    'list-ordered',
    'list-unordered',
  ]),
  addClass: PropTypes.string,
};

export default List;
