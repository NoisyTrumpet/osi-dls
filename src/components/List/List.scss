.list-horizontal {
  display: flex;
}

.list-ordered,
.list-unordered {
  margin-bottom: $space-12;

  > li {
    margin-top: $space-8;
    margin-left: $space-16;
  }
}

.list-ordered {
  list-style-type: decimal;

  // .list-ordered {
  //   list-style-type: lower-latin;
  // }
}

.list-unordered {
  list-style-type: disc;

  // .list-ordered {
  //   list-style-type: circle;
  // }
}

ul.list-ordered {
  @include state-does-not-exist('Unordered list cannot be ordered');
}
