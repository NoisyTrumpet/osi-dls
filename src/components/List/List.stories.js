import React from "react";
import { storiesOf } from "@storybook/react";
import { withSmartKnobs } from "storybook-addon-smart-knobs";
import { List } from "./";
import { Typography } from "../Typography";

const storyTitle = "components/List"; // Stories can be organized in a nested structure using ”/” as a separator, and can be given a top-level heading using a ”|” root separator

const parameters = {
  component: List,
  componentSubtitle:
    "The <List /> tag defines a list component (defaults to an unordered list)",
  jest: ["List"]
};

storiesOf(storyTitle, module)
  .addParameters(parameters)
  .addDecorator(withSmartKnobs({ ignoreProps: ["children"] }))
  .add("Default", () => (
    <List>
      <List.Item>
        <Typography variant="body">Text 1</Typography>
      </List.Item>
      <List.Item>
        <Typography variant="body">Text 2</Typography>
      </List.Item>
      <List.Item>
        <Typography variant="body">Text 3</Typography>
      </List.Item>
    </List>
  ))
  .add("List within a list", () => (
    <List>
      <List.Item>
        <Typography variant="body">Text 1</Typography>
      </List.Item>
      <List.Item>
        <Typography variant="body">Text 2</Typography>
      </List.Item>
      <List.Item>
        <Typography variant="body">Text 3</Typography>
        <List>
          <List.Item>
            <Typography variant="body">Text 4</Typography>
          </List.Item>
          <List.Item>
            <Typography variant="body">Text 5</Typography>
          </List.Item>
          <List.Item>
            <Typography variant="body">Text 6</Typography>
          </List.Item>
        </List>
      </List.Item>
    </List>
  ));
