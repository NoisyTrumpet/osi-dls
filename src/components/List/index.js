export { default as List } from './List'; // eslint-disable-line import/prefer-default-export
