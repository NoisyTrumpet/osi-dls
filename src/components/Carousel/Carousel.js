import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import GliderJS from './GliderJS/glider';
import { List } from '../List';
import './Carousel.scss';
import { canUseDOM } from '../../util';

/**
 * A configurable carousel that utilizes Glider.js (https://nickpiscitelli.github.io/Glider.js/)
 */
const Carousel = ({
  children,
  addClass,
  slidesToShow,
  dots,
  arrows,
  scrollLockDelay,
  duration,
  itemWidth,
  enableRewindLooping,
  disableDragging,
  enableScrollLock,
  onNewItemVisible,
  squared,
}) => {
  const carouselRef = useRef(null);
  const gliderRef = useRef(null);
  const [loaded, setLoaded] = useState(false); // Flagged true after GliderJS inits
  const gliderQuerySelector = '.glider';

  const classes = classNames(
    'carousel',
    { 'carousel-square': squared }, // 1:1 Aspect ratio
    addClass,
  );

  // Options for GliderJS (https://nickpiscitelli.github.io/Glider.js/)
  const createOptionsObject = () => ({
    // configurable options
    rewind: enableRewindLooping,
    draggable: !disableDragging,
    scrollLock: enableScrollLock,
    slidesToShow: itemWidth ? 'auto' : slidesToShow,
    itemWidth,
    // 'undefined' is glider.js's default value for exactWidth
    exactWidth: itemWidth ? true : undefined,
    scrollLockDelay,
    duration,
  });

  // Called whenever a slide is visible
  const itemVisible = (event) => {
    if (onNewItemVisible) {
      const itemIndex = event.detail.slide;
      onNewItemVisible(itemIndex);
    }
  };

  // Called whenever a slide is hidden
  const itemHidden = (e) => {
    const slideIndex = e.detail.slide;
    const hiddenSlide = carouselRef.current.querySelectorAll('.glider-slide')[slideIndex];
    // Assumes there is only 1 video per slide
    if (hiddenSlide.querySelector('video')) {
      const video = hiddenSlide.querySelector('video');
      video.pause();
    }
  };

  // On mount, init GliderJS and add event listeners (when window is not undefined)
  useEffect(() => {
    if (canUseDOM() && !gliderRef.current) {
      GliderJS(); // Runs GliderJS code found in src/components/Carousel/GliderJS/glider.js
      const carousel = carouselRef.current;
      gliderRef.current = new Glider(carousel.querySelector(gliderQuerySelector), { // eslint-disable-line
        ...createOptionsObject(),
        /** Selectors */
        dots,
        arrows,
        /** Do not change */
        skipTrack: true,
        resizeLock: true,
      });
      carousel.querySelector(gliderQuerySelector).addEventListener('glider-slide-hidden', itemHidden);
      carousel.querySelector(gliderQuerySelector).addEventListener('glider-slide-visible', itemVisible);
      setLoaded(true);
    }
    return () => {
      const carousel = carouselRef.current;
      carousel.querySelector(gliderQuerySelector).removeEventListener('glider-slide-hidden', itemHidden);
      carousel.querySelector(gliderQuerySelector).removeEventListener('glider-slide-visible', itemVisible);
      gliderRef.current.destroy();
    };
  }, [window]);

  return (
    <div ref={carouselRef} className={classes}>
      <div className="glider carousel__track-container">
        <List addClass="carousel__track glider-track">
          {React.Children.map(children, (child) => (<List.Item>{loaded && child}</List.Item>))}
        </List>
      </div>
    </div>
  );
};

Carousel.defaultProps = {
  slidesToShow: 1,
  scrollLockDelay: 100,
  duration: 0.5,
};

Carousel.propTypes = {
  /** Carousel items */
  children: PropTypes.node,
  /** Class to append to <Carousel />  */
  addClass: PropTypes.string,
  /** Callback function when a new item is visible */
  onNewItemVisible: PropTypes.func,
  /** The number of items to show at once.
    * Set the 'itemWidth; prop instead if you prefer to show items based off their width. */
  slidesToShow: PropTypes.number,
  /** If true, the Carousel ignores dragging with the mouse */
  disableDragging: PropTypes.bool,
  /** When scrollLock is set, this is the delay in milliseconds to wait before the scroll happens */
  scrollLockDelay: PropTypes.number,
  /** If true, Carousel will scroll to the beginning/end when it's respective endpoint is reached */
  enableRewindLooping: PropTypes.bool,
  /** If true, Carousel will scroll to the nearest slide after any scroll interactions */
  enableScrollLock: PropTypes.bool,
  /** If true, Carousel will be have a 1:1 aspect ratio */
  squared: PropTypes.bool,
  /** An aggravator used to control animation speed. Higher is slower. */
  duration: PropTypes.number,
  /** Setting this prop will cause the 'slidesToShow' prop to be ignored. */
  itemWidth: PropTypes.number || PropTypes.func,
  /** An HTML selector containing the dot container */
  dots: PropTypes.string,
  /** An HTML selector for the arrow buttons */
  arrows: PropTypes.exact({
    prev: PropTypes.string,
    next: PropTypes.string,
  }),
};

export default Carousel;
