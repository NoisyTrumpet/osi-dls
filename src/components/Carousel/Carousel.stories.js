import React from "react";
import { storiesOf } from "@storybook/react";
import { withSmartKnobs } from "storybook-addon-smart-knobs";
import Carousel from "./Carousel";
import { Typography } from "../Typography";

const storyTitle = "components | Carousel"; // Stories can be organized in a nested structure using ”/” as a separator, and can be given a top-level heading using a ”|” root separator

const parameters = {
  component: Carousel,
  componentSubtitle: "<Carousel /> defines a Carousel",
  jest: ["Carousel"]
};

const exampleStyles = {
  productCarousel: {
    item: {
      height: "300px",
      width: "100%"
    }
  },
  signAndSend: {
    outer: {
      height: "100%",
      width: "100%",
      boxSizing: "border-box",
      padding: "12px"
    },
    inner: {
      height: "100%",
      width: "100%",
      borderRadius: "3px",
      boxShadow: "1px 1px 10px lightgrey",
      background: "white"
    }
  }
};

// Pictures from https://osilife.com/cards/greeting-cards/assorted-cards-for-all-occasions-in-floral-organizer-box-box-of-24-1BCE4202.html
storiesOf(storyTitle, module)
  .addParameters(parameters)
  .addDecorator(
    withSmartKnobs({
      ignoreProps: [
        "children",
        "dots",
        "arrows",
        "enableScrollLock",
        "slidesToShow",
        "squared"
      ]
    })
  )
  .add("Product Carousel (fixed item width)", () => (
    <div style={{ padding: "32px", background: "white" }}>
      <Typography variant="headline-5" color="gray5">
        Title
      </Typography>
      <Carousel itemWidth={210} enableScrollLock={false}>
        <div
          style={{ background: "#aaa", ...exampleStyles.productCarousel.item }}
        />
        <div
          style={{ background: "#ccc", ...exampleStyles.productCarousel.item }}
        />
        <div
          style={{ background: "#aaa", ...exampleStyles.productCarousel.item }}
        />
        <div
          style={{ background: "#ccc", ...exampleStyles.productCarousel.item }}
        />
        <div
          style={{ background: "#aaa", ...exampleStyles.productCarousel.item }}
        />
        <div
          style={{ background: "#ccc", ...exampleStyles.productCarousel.item }}
        />
        <div
          style={{ background: "#aaa", ...exampleStyles.productCarousel.item }}
        />
      </Carousel>
    </div>
  ))
  .add("Sign and Send (Dot controls)", () => (
    <div style={{ width: "272px", height: "404px", background: "grey" }}>
      <Carousel dots="#example-dots" enableScrollLock>
        <div style={exampleStyles.signAndSend.outer}>
          <div style={exampleStyles.signAndSend.inner} />
        </div>
        <div style={exampleStyles.signAndSend.outer}>
          <div style={exampleStyles.signAndSend.inner} />
        </div>
        <div style={exampleStyles.signAndSend.outer}>
          <div style={exampleStyles.signAndSend.inner} />
        </div>
      </Carousel>
      <div role="tablist" id="example-dots" />
    </div>
  ));
