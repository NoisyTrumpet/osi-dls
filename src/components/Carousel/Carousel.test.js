import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Carousel } from '../Carousel';
import { List } from '../List';

describe('<Carousel />', () => {
  test('should render', () => {
    const wrapper = shallow(<Carousel />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should render with correct number of children', () => {
    const wrapper = shallow(
      <Carousel>
        <div />
        <div />
        <div />
      </Carousel>,
    );
    expect(wrapper.find(List.Item).length).toBe(3);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
