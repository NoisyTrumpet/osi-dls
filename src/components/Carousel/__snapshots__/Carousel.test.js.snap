// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> should render 1`] = `
<div
  className="carousel"
>
  <div
    className="glider carousel__track-container"
  >
    <List
      addClass="carousel__track glider-track"
    />
  </div>
</div>
`;

exports[`<Carousel /> should render with correct number of children 1`] = `
<div
  className="carousel"
>
  <div
    className="glider carousel__track-container"
  >
    <List
      addClass="carousel__track glider-track"
    >
      <ListItem
        key=".0"
      />
      <ListItem
        key=".1"
      />
      <ListItem
        key=".2"
      />
    </List>
  </div>
</div>
`;
